### ISSUE TITLE
when there's a delay between additional rows of a table being output, they are printed as separate tables

### ISSUE THREAD

**Issue by BLucky-gh** [nushell/nushell#9405]
### Describe the bug

When filtering through a large table, or when doing an `each` with a closure that takes a bit of time, or any other situation that produces a delay between rows being output, they get split into separate tables
here are some examples:

an `each` doing http requests that each takes about 5 secs to complete:
![image](https://github.com/nushell/nushell/assets/41377306/e8ed0caa-3a62-4daa-aca9-d8dd11406c97)
vs an `each` just echoing the input back immediately:
![image](https://github.com/nushell/nushell/assets/41377306/eba234c8-c087-47f6-8ae5-a05a4f823154)

I can `collect` the results of the former with a dumb closure to produce a single table, but that is a hack:
![image](https://github.com/nushell/nushell/assets/41377306/55281ac2-85f0-4c12-a1f4-0afc359b0136)

another example, filtering through a large sqlite database:
![image](https://github.com/nushell/nushell/assets/41377306/221b996e-e113-4d0d-8a9b-3ec93321beef)

`1..10 | each { || sleep 1sec; echo done}`
![image](https://github.com/nushell/nushell/assets/41377306/230d6f87-261b-4c9c-b97f-42e77e20ed27)

`1..10 | each { || sleep 0.5sec; echo done}`
![image](https://github.com/nushell/nushell/assets/41377306/6a7de107-3b82-469c-b559-28d1f9266463)

`1..10 | each { || sleep 0.25sec; echo done}`
![image](https://github.com/nushell/nushell/assets/41377306/3a576a80-ed7f-4693-8088-860f8fc55f8e)

(looks like the tables are closed every second)

### How to reproduce

`1..10 | each { || sleep 1sec; echo done}`

### Expected behavior

The tables should be one big table 

### Configuration

| key                | value                               |
| ------------------ | ----------------------------------- |
| version            | 0.80.0                              |
| branch             |                                     |
| commit_hash        |                                     |
| build_os           | macos-x86_64                        |
| build_target       | x86_64-apple-darwin                 |
| rust_version       | rustc 1.68.2 (9eb3afe9e 2023-03-27) |
| rust_channel       | stable-x86_64-apple-darwin          |
| cargo_version      | cargo 1.68.2 (6feb7c9cf 2023-03-26) |
| build_time         | 2023-06-04 22:24:58 +04:00          |
| build_rust_channel | release                             |
| features           | default, zip                        |
| installed_plugins  |                                     |

note: I also compiled it with dataframes support, which for some reason isn't showing here. In case it's relevant the exact command I installed `nu` with is `RUSTFLAGS=-Ctarget-cpu=native cargo install nu --features=dataframe --config profile.release.opt-level=3  --config profile.release.package.'"*"'.opt-level=3 --config profile.release.lto='"fat"' --config profile.release.codegen-units=1 --config profile.release.strip=true`

Except for the dataframe support, the extra flags were just there to squeeze as much marginal performance out of it as possible because I simply have nothing better to do.

**Comment by amtoine** [comment]
yeah i think i've seen this already :thinking:

**Comment by fdncred** [comment]
yes, this is by design. nushell tables get generated in batches based on a timeout or 1000 rows. this is to prevent the appearance of the ui hanging. one way around that is adding `collect` to your pipeline. e.g.
```
❯ 1..10 | each {|| sleep 250ms; echo done} | collect {|r| $r}
```
or
```
❯ 1..10 | each {|| sleep 250ms; echo done} | collect { $in }
```

**Comment by BLucky-gh** [comment]
while avoiding it looking like the UI is hanging is a good reason, wouldn't it make sense to append to the table that's already been printed without printing the bottom border of the table rather than make it look like another table? it can still print out a couple rows at a time, just not as an entirely new table.

**Comment by fdncred** [comment]
So, we have to know if this is the last row, in order to draw the last row separators, if we assume every row is not the last row, we won't draw the table right. Since we can stream tables infinitely, we'd have to have a better algorithm that has a spinner to indicate that we're waiting on another row, or something like that. I really don't like batching the table output by timeout and 1000s myself, so I'd love to have something better.

**Comment by BLucky-gh** [comment]
I'm not sure how the nushell functions run in underlying rust code, but can't that be handled by something akin to an `EOF` signal in filesystems? Considering that the batching stuff is a problem in the first place I assume that's not how it's implemented, but since only `nu` internal programs can print records and tables, it is all handled by the underlying rust code of nushell, tho I guess implementing this now is non-trivial.

or if a given program can only return a single record/table at a time (which is most probably the case), one could also just determine the start and end of the record from when the program returning it begins and ends.

**Comment by fdncred** [comment]
We'd be open to PRs exploring different ways to tweak this algorithm.
